File: cpp/fft.cpp
#include <chrono>
#include <complex>
#include <iostream>
#include <functional>
#include <numbers>
#include <stdexcept>
#include <string>
#include <vector>

#include "vector_func.cpp"

enum MODE { BCK, RUN };

void benchmark(int runs, int num, std::function<std::vector<std::complex<double>>(std::vector<std::complex<double>>)> func)
{
    auto x = random_array(num);
    std::vector<float> times(runs);
    using FpSeconds = std::chrono::duration<float, std::chrono::seconds::period>;

    static_assert(std::chrono::treat_as_floating_point<FpSeconds::rep>::value,
                  "Rep required to be floating point");

    for (int i=0; i < runs; i++) {
        auto start = std::chrono::high_resolution_clock::now();
        func(x);
        auto stop = std::chrono::high_resolution_clock::now();
        times[i] = FpSeconds(stop - start).count();
    }

    std::cout << "Run took: " << mean_vector(times) << " ± " << std_vector(times) << " s" << std::endl;

}

std::vector<std::complex<double>> dft(std::vector<std::complex<double>> const &x)
{
    int N = x.size();
    std::vector<std::complex<double>> X(N);
    std::complex<double> I(0.0, 1.0);
    std::complex<double> w = std::exp(-2.0 * I * std::numbers::pi / (double)N);;

    for (int j=0; j < N; j++) {
        std::complex<double> ss(0.0, 0.0);
        for (int k=0; k < N; k++) {
            ss += x[k] * std::pow(w, k*j);
        }
        X[j] = ss;
    }
    return X;
}

std::vector<std::complex<double>> fft(std::vector<std::complex<double>> const &x)
{
    int N = x.size();
    std::complex<double> I(0.0, 1.0);
    if (!((N & (N - 1)) == 0)) {
        throw std::invalid_argument("Number of elements must be a power of 2");
    }
    else if (N < 16) {
        return dft(x);
    }
    else {
        auto X_even = fft(get_even(x));
        auto X_odd = fft(get_odd(x));

        std::vector<std::complex<double>> factor(N);
        for (int n=0; n < N; n++) {
            factor[n] = std::exp(-2.0 * std::numbers::pi * I * (double)  n / (double) N);
        }

        auto sum1 = mult_vector(slice_vector(factor, 0, (N / 2) - 1), X_odd);
        auto first = add_vector(X_even, sum1);

        auto sum2 = mult_vector(slice_vector(factor, N / 2, N - 1), X_odd);
        auto last = add_vector(X_even, sum2);

        first.insert(first.end(), last.begin(), last.end());
        return first;
    }

}

int main(int argc, char** argv)
{

    MODE mode = RUN;
    int num, reps;
    std::function<std::vector<std::complex<double>>(std::vector<std::complex<double>>)> func;

    if (argc > 1) {

        std::string arg(argv[1]);
        if (arg == "FFT") {
            func = fft;
        }
        else if (arg == "DFT") {
            func = dft;
        }
        else {
            throw std::invalid_argument("Must pass a valid algorithm name (FFT, DFT)");
        }
        num = std::stoi(argv[2]);

    }
    if (argc > 3) {
        std::string arg(argv[3]);
        if (arg == "benchmark") mode = BCK;
        reps = std::stoi(argv[4]);

    }

    switch (mode)
    {
    case BCK:
        std::cout << "Benchmarking" << std::endl;
        benchmark(reps, num, func);
        break;

    case RUN:
        auto x = random_array(num);
        func(x);
    }
    return 0;

}

File: python/fft.py
import sys
import time

import numpy as np

BCK, RUN = 0, 1

def random_array(num):
    return np.random.rand(num)*2.0-1.0 + 1j*np.random.rand(num)*2.0-1.0

def benchmark(runs, num, func):

    x = random_array(num)
    times = np.empty(runs)

    for i in range(runs):
        start = time.perf_counter()
        func(x)
        stop = time.perf_counter()
        times[i] = stop - start

    print(f"Run took: {times.mean()} ± {times.std()} s")

def fft_np(x):
    return np.fft.fft(x)

def dft_vec(x):
    N = x.shape[0]
    n = np.arange(N)
    k = n.reshape((N, 1))
    W = np.exp(-2j * np.pi * k * n / N)
    return W @ x

def fft(x):
    N = x.shape[0]
    if not ((N & (N - 1)) == 0):
        raise ValueError("Number of elements must be a power of 2")
    elif N <= 16:
        return dft_vec(x)
    else:
        X_even = fft(x[::2])
        X_odd = fft(x[1::2])
        n = np.arange(N)
        factor = np.exp(-2j*np.pi*n/N)
        return np.concatenate([X_even + factor[:N // 2] * X_odd,
                               X_even + factor[N // 2:] * X_odd])

def dft(x):
    N = x.shape[0]
    X = np.zeros(N, dtype=complex)

    for j in range(N):
        ss = 0.0 + 1j*0.0
        for k in range(N):
            w = np.exp(-j*k*2j*np.pi/N)
            ss += x[k]*w
        X[j] = ss
    return X

if __name__ == "__main__":

    mode = RUN
    num, reps = 1024, 100
    func = fft

    if len(sys.argv) > 1:

        arg = sys.argv[1]
        if arg == "FFT":
            func = fft
        elif arg == "DFT-VEC":
            func = dft_vec
        elif arg == "DFT":
            func = dft
        elif arg == "FFT-NP":
            func = fft_np
        num = int(sys.argv[2])

    if len(sys.argv) > 3:
        arg = sys.argv[3]
        if arg =="benchmark": mode = BCK
        reps = int(sys.argv[4])

    if mode == BCK:
        print("Benchmarking")
        benchmark(reps, num, func)

    elif mode == RUN:
        x = random_array(num)
        func(x)